libs/hpc-ui/src/lib/components/form-fields/upload-file-field.tsx

```diff
@@ -0,0 +1,239 @@
+import ChangeCircleIcon from '@mui/icons-material/ChangeCircle';
+import DeleteIcon from '@mui/icons-material/Delete';
+import FileDownloadIcon from '@mui/icons-material/FileDownload';
+import FilePresentIcon from '@mui/icons-material/FilePresent';
+import { Box } from '@mui/material';
+import { styled } from '@mui/material/styles';
+import { type util } from '@unocha/hpc-data';
+import { useRef, useState } from 'react';
+import tw from 'twin.macro';
+import { type LanguageKey, t } from '../../i18n';
+import AsyncIconButton from '../async-icon-button';
+import { Button, type ButtonProps } from '../button';
+
+type UploadFileProps = {
+  name: string;
+  buttonConfig: Omit<ButtonProps, 'onClick'>;
+  onUpload: (file?: File) => Promise<unknown>;
+  onDelete?: (
+    setSavedFile: React.Dispatch<React.SetStateAction<File | undefined>>
+  ) => Promise<unknown>;
+  onDownload?: {
+    handleDownload: () => Promise<unknown>;
+    handleErrorToast?: (error: Error) => void;
+  };
+  file?: util.FormObjectValue;
+  confirmUpload?: {
+    validation: (file?: File, ...args: unknown[]) => unknown;
+    onSuccess: (fileName: string) => unknown;
+    onError: (err: unknown) => unknown;
+  };
+  disabled?: boolean;
+  lang?: LanguageKey;
+  hideFileChangeStatusStyle?: boolean;
+  validFileFormats?: {
+    whitelist: Set<string>;
+    validationErrorToast: () => unknown;
+  };
+};
+
+/**
+ * These styles come from:
+ * https://mui.com/material-ui/react-button/#file-upload
+ */
+const VisuallyHiddenInput = styled('input')({
+  clip: 'rect(0 0 0 0)',
+  clipPath: 'inset(50%)',
+  height: 1,
+  overflow: 'hidden',
+  position: 'absolute',
+  bottom: 0,
+  left: 0,
+  whiteSpace: 'nowrap',
+  width: 1,
+});
+
+const FileViewContainer = tw.div`
+  flex
+  gap-x-10
+  justify-around
+  items-center
+  p-4
+  my-4
+  border
+  border-solid
+  border-unocha-panel-border
+  rounded-sm
+`;
+
+const ButtonsContainer = tw.div`
+  flex
+  flex-col
+  items-center
+`;
+
+const OverflowSpan = tw.span`
+  break-all
+  overflow-hidden
+`;
+
+const validateExtension = (filename: string, whitelist: Set<string>) => {
+  const parts = filename.split('.');
+  if (!parts.length) {
+    return '';
+  }
+  const extension = parts.at(-1)?.toLowerCase() ?? '';
+  return whitelist.has(extension);
+};
```

Comment:
Your return type is `boolean | ""`. Why return empty string and not `false`. The real problem is that your function returns `true` for this: `validateExtension('png', WHITELIST_FILE_EXTENSIONS);`. Let's do this instead:

```suggestion
const validateExtension = (filename: string, whitelist: Set<string>) => {
  const lastDot = filename.lastIndexOf('.');
  if (lastDot <= 0) {
    return false;
  }
  const extension = filename.slice(lastDot + 1).toLowerCase();
  return whitelist.has(extension);
};
```